New stretch script

electrocbd opened the pull request:
This new script performs "post stretch" algorithm to fix the problem of insufficient inner and outer diameters

electrocbd committed:
New stretch script

electrocbd committed:
Security when first and last point of a triangle are too close during
computing of WideTurn and WideCircle

nallath commented:
for our reference; CURA-3781

nallath commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,339 @@
+# Copyright (c) 2015 Jaime van Kessel, Ultimaker B.V.
```
I don't think I have any copyrights over this ;)

nallath commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,339 @@
+                    "minimum_value_warning": 0,
+                    "maximum_value_warning": 0.2
+                },
+                "line_width":
```
isn't this a setting that you need to get from the active machine?

electrocbd replied:
Yes it is. Is there any way to get it?

nallath replied:
ExtruderManager.getInstance().getActiveExtruderStack() .getProperty("line_width", "value") if it's a multi extrusion machine or Application.getInstance().getGlobalContainerStack().getProperty("line_width", "value")

nallath commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,339 @@
+        self.line_width = self.getSettingValueByKey("line_width")
+        self.stretch = self.getSettingValueByKey("stretch")
+        retdata = []
+        LayerSteps = []
```
Codestyle; atributes should be "layer_steps"

nallath commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,339 @@
+        }"""
+
+    def execute(self, data):
+        self.line_width = self.getSettingValueByKey("line_width")
```
codestyle; atributes in self should be defined in init

electrocbd replied:
There is a problem, the initial value is defined after init is called. It is a "self" value because it can be different between instances of stretch objects, so what to do? Define it empty in init and overwrite it in execute?

electrocbd replied:
I split class into execution (Stretcher) and setup (Stretch) to allow initialization of attributes during construction

nallath commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,339 @@
+            lines = layer.rstrip("\n").split("\n")
+            for line in lines:
+                comment = ""
+                if line.find(';') >= 0:
```
Codestyle: use " instead of '

nallath commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,339 @@
+                else:
+                    onestep = GCodeStep(-1,current_x,current_y,current_z,current_e,current_f,line);
+                if current_z != layer_z:
+                    print("Layer Z " + "{:.3f}".format(layer_z) + " " + str(len(LayerSteps)) + " steps")
```
If you want to print debug info, please use Logger.log("d", "message")

nallath commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,339 @@
+        if len(LayerSteps):
+            retdata.append(self.ProcessLayer(LayerSteps))
+        return retdata
+    def ProcessLayer(self,LayerSteps):
```
Add empty line (functions need to have ampty line between them)

nallath commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,339 @@
+        curE = LayerSteps[0].e
+        vPos = np.empty((0,2))
+        iflush = 0
+        for i in range(len(LayerSteps)):
```
You can use enumerate for this. Example; for i, step in enumerate(layer_steps):
This gives you both acces to the current step and the index.

electrocbd replied:
Good tip :-)

nallath commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,339 @@
+        iflush = 0
+        for i in range(len(LayerSteps)):
+            if i==0:
+                curE = LayerSteps[i].e
```
codestyle; should be current_e

nallath commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,339 @@
+            self.vd1 = np.concatenate([self.vd1,np.array(vPos[:-1])])
+            self.vd2 = np.concatenate([self.vd2,np.array(vPos[1:])])
+
+    def WideCircle(self,vPos,vTrans):
```
Codestyle; functions should be lowerCamalCase

electrocbd committed:
stretch plugin copyright

electrocbd committed:
Coding style
Use of UM.Logger

electrocbd committed:
Split into execution class Stretcher and setup class Stretch

electrocbd committed:
Coding style, use double quotes

electrocbd committed:
Coding style

electrocbd commented:
Ready for a new review :-)

Ghostkeeper commented:
What's wrong with the Horizontal Expansion setting for this purpose? Or Outer Wall Inset?

BagelOrb commented:
@ghostkeeper
Inner corners should be handled differently from outer corners.

Inner holes are typically smaller than the mesh dimensions compared to
outside contours however you tweak horizontal expansion.

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -39,10 +39,10 @@ def __init__(self,line_width,stretch):
             return default
         sub_part = line[line.find(key) + 1:]
-        m = re.search('^-?[0-9]+\.?[0-9]*', sub_part)
+        m = re.search("^-?[0-9]+\.?[0-9]*", sub_part)
```
The `\.` is going to be unescaped in Python to `.` which then means "any symbol".

The most common way around that is to put an `r` before the regular expression, like so:

    m = re.search(r"^-?[0-9]+\.?[0-9]*", sub_part)

electrocbd replied:
Done. I suggest doing the same in ..Script.py :-)

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,357 @@
+                    onestep = GCodeStep(-1,current_x,current_y,current_z,current_e,current_f,comment)
+                else:
+                    onestep = GCodeStep(-1,current_x,current_y,current_z,current_e,current_f,line);
+                if current_z != layer_z:
```
This is going to break when Z-hops are activated!

And spiralize is right out, but I don't suppose you intend to support that.

electrocbd replied:
The main objective is to be able to determine for any segment if material is already deposited at any side of the segment at the same height (not for the previous layer). I don't know it it makes any sense in spiralize mode.
For z-hops, the easiest solution should be to take into account the information coming from curaengine, but I noticed that the array provided to the plugin is not synchronized with layers (several layers can be in the same index)

electrocbd replied:
Anyway, as you have certainly noticed, I have a poor knowledge of python and did a rough translation from French C++ into English Python. I will try to enhance ;-)

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,357 @@
+        current_e = layerSteps[0].e
+        vPos = np.empty((0,2))
+        iflush = 0
+        for i in range(len(layerSteps)):
```
Python has a function for this:

    for i, step in enumerate(layerSteps):
        #Now i has the iteration index and step is the value of layerSteps[i].

Not really important, but I thought it's useful to know if you come from a background of Java or C++. Enumerate is a function that returns an iterator over tuples, and the `i, step` automatically unfolds those tuples for you into separate variables to use.

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,357 @@
+        for i in range(len(layerSteps)):
+            if i==0:
+                current_e = layerSteps[i].e
+            if current_e == layerSteps[i].e:
```
Could you document this? It looks like this `if` statement is entered when a step didn't introduce any `E`, i.e. it's a travel move.

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,357 @@
+        if len(vPos):
+            vTrans = np.copy(vPos)
+        if len(vPos)>=2:
+            self.workOnSequence(vPos,vTrans)
```
It looks like every time you call workOnSequence, you always just copied vPos into vTrans. Isn't it better to just provide workOnSequence with one copy of the array and do copying inside workOnSequence if necessary? You can have it return the resulting vTrans. I think that makes the function more intuitive.

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,357 @@
+        if onestep.f != self.output_f:
+            self.output_f = onestep.f
+            sout += " F"
+            sout += "{:.0f}".format(self.output_f).rstrip(".")
```
Might as well put the `" F"` in the formatting string :wink:

    sout += " F{:.0f}".format(self.output_f).rstrip(".")

electrocbd replied:
Done

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,357 @@
+            sout += " F"
+            sout += "{:.0f}".format(self.output_f).rstrip(".")
+        if onestep.x != self.output_x or onestep.y != self.output_y or onestep.z != self.output_z:
+            assert onestep.x >= 0 and onestep.x < 200 # Security
```
You're going to need to remove these asserts, since they don't hold for printers larger than 20cm. Lots of printers are larger than 20cm.

electrocbd replied:
Done. I left an assert for really insane huge printers, just in case...

Ghostkeeper replied:
It also doesn't work for printers which use `machine_center_is_zero`, because those may give negative coordinates ;)

electrocbd replied:
The assert is now between -1000 and +1000. I did a google search to look for FDM printers more than 1 meter wide, and didn't find any ;-)
My first version of the plugin (more precisely, the C++ version)  didn't have any assertion, but for a specific test case (some kind of divide by zero), now fixed, the modified gcode asked the instrument to go to position NaN. The problem is that Ultimaker2 tried to go there, and bumped to end limits repeatedly. I don't think that this is good for the mechanics...

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,357 @@
+        return sout
+
+    def generate(self,layerSteps,i,iend,vPos,layergcode):
+        ipos = 0
```
What's the difference between `i`, `ipos` and `iend`?

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,357 @@
+        The first and last point of the sequence are the same,
+        so it is possible to extend the end of the sequence with its beginning when seeking for triangles
+        '''
+        d1 = self.line_width / 2.0
```
In general, the code is easier to understand if you use words for variable names rather than letters and numbers. `d1` is half a line width, apparently. `d4` is a stretch factor (why 4?), `i1`, `i` and `i3` are mysteries to me!

electrocbd replied:
Ok, I made a couple of hours making pylint happy. I hope my code is more readable now :-)

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,357 @@
+        while i<len(vPos):
+            vPosAfter = np.resize(np.roll(vPos,-i-1,0),(iextra,2))
+            good_triangle = True
+            dp = ((vPos[i]-vPosAfter)**2).sum(1)
```
Numpy has built-in functions for dot products.

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,357 @@
+            good_triangle = True
+            dp = ((vPos[i]-vPosAfter)**2).sum(1)
+            if np.amax(dp) < d1*d1:
+                good_triangle = False
```
You don't need this boolean here. You can just use `continue` as keyword.

electrocbd replied:
Done.

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,357 @@
+        i1 = 0
+        i = 0
+        i3 = 0
+        while i<len(vPos):
```
A for-each loop seems appropriate here, such as

    for i in range(len(vPos)):

Or if possible:

    for i, vertex in enumerate(vPos):

The length of vPos is not going to change.

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,357 @@
+                pp=vPosBefore[i1] + r*(vPosAfter[i3]-vPosBefore[i1])
+                dpp=np.sqrt(((pp-vPos[i])**2).sum(0))
+                if dpp > 0.001:
+                    vTrans[i] = vPos[i] - (d4/dpp)*(pp-vPos[i])
```
What's all this trying to compute in line 211 to 221? I see a lot of Pythagoras going on and something that looks like the actual move of the point. Name your variables, please!

electrocbd replied:
I added comments.

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,357 @@
+        i1 = 0
+        i = 1
+        i3 = 2
+        while i+1<len(vPos):
```
Again, this is a good candidate for a `for` loop. And the same applies too about the `continue` keyword, the dot product and the variable names.

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,357 @@
+            if i2 == len(vPos):
+                i2 = i-1
+            xm = vPos[i1]
+            xperp = np.dot(mrot,vPos[i2]-vPos[i1])
```
So this rotates the direction vector after `i` 90 degrees counter-clockwise. The resulting direction vector is scaled and then added to the vector at `i`. The problem is then that the direction vector before `i` is not taken into account. If the angle is more than 90 degrees, then this would result in the vertex being moved towards the new line, making a smaller turn rather than a larger one. To properly produce the normal vector of a vertex, you need to take both lines into account. It's commonly done by taking the average of the two direction vectors going away from the vertex (and guarding against zero-vectors then). That's also much simpler than computing a perpendicular.

electrocbd replied:
The objective of this computation is not to move the point, but to find for the current segment if material was already deposited at one side or at the other side of this segment. To test if material is present, I look at the position from the first point of the segment (it was initially the middle, but this doesn't have a big impact), in the perpendicular direction, at a distance of the line size. If one of the segments of already deposited material is near the test point, there is material at this side.

This is the slowest part of the algorithm. It takes few seconds in C++, much more in python.

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,357 @@
+            xperp = np.dot(mrot,vPos[i2]-vPos[i1])
+            xperp = xperp / np.sqrt((xperp**2).sum(-1))
+            xp1 = xm + xperp * d2
+            toucheplus = False
```
Please write your code in English! Well, in Python-English. Parseltongue? Parsertongue?

Ghostkeeper reviewed (changes requested):
I'm sorry to be so critical. A lot of the code is quite unreadable though. The main reasons for that are the usage of single-letter variable names and the general lack of documentation. I did my best but I can't say that I fully understand everything, because of that. I made some suggestions to improve on that.

electrocbd committed:
Allow printing area greater than 20cm x 20cm

electrocbd committed:
Coding style (spaces)

electrocbd committed:
Coding style

electrocbd committed:
Coding style

electrocbd committed:
Coding style

electrocbd committed:
Comments

electrocbd committed:
Inefficient coding

electrocbd committed:
Fix name of private module function
Fix wrong attribute name

electrocbd committed:
Refactoring, methods moved into GCode class

electrocbd commented:
I finished my rework. I added comments, named my variables, etc. It should be more readable
The plugin seems to work as before

LipuFei commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,419 @@
+Python implementation of https://github.com/electrocbd/post_stretch
+Correction of hole sizes, cylinder diameters and curves
+
+WARNING This script has never been tested with several extruders
```
Could you add a general description of what this script does? I cannot find this information anywhere...

Ghostkeeper replied:
It says:

> Correction of hole sizes, cylinder diameters and curves.

See the original description in https://github.com/electrocbd/post_stretch

LipuFei commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,419 @@
+from UM.Application import Application
+import re
+
+def _getValue(line, key, default=None):
```
I think a name like `_getValueFromGcodeCommand()` would be more descriptive.

LipuFei replied:
ok... so this function is used to get a gcode command number ("G92" for example) or a value in a command (like "X12" in "G0 X12"). I thought it is only used for things like "X12". Could you update the doc as well?

nallath replied:
I think it's quite like the getValue function already in the script.py

LipuFei commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,419 @@
+    I split the class into setup part (Stretch) and execution part (Strecher)
+    and only the setup part inherits from Script
+    """
+    if not key in line or (";" in line and line.find(key) > line.find(";")):
```
~I think this is not safe. At least exclude the Gcode command first when searching for the key. could be a preprocess like `line = " ".join(line.strip().split(" ")[1:])`~

Sorry that I missed the part that this is also used for commands like "G92". this change is not needed.

LipuFei commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,419 @@
+        """
+        Computes the new X and Y coordinates of all g-code steps
+        """
+        Logger.log("d", "Post stretch with line width=" + str(self.line_width)
```
You can use something like:
```
Logger.log("d", "Post stretch with line width = %smm and stretch = %smm", self.line_width, self.stretch)
```

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -165,25 +167,25 @@ def stepToGcode(self, onestep):
+            sout += " Z{:.3f}".format(self.outpos.step_z).rstrip("0").rstrip(".")
+        if onestep.step_e != self.outpos.step_e:
+            self.outpos.step_e = onestep.step_e
+            sout += " E{:.5f}".format(self.outpos.step_e).rstrip("0").rstrip(".")
```
Going object-oriented, this function might as well be a function of the GCodeStep class now. Just a style-thing though, if you agree.

(Not blocking.)

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,419 @@
+                    layer_z = current.step_z
+                layer_steps.append(onestep)
+        if len(layer_steps):
+            retdata.append(self.processLayer(layer_steps))
```
If you indent this one step further, then the entire clause above of `if current.step_z != layer_z:` would be unnecessary. You already get the layers separated, so there's no need to separate them again by Z coordinate. Separating them by Z coordinate is also incorrect if there are Z-hops, spiralised models, extruder switches or minimum-layer-time-constrained layers. Those change the Z coordinate within a layer.

If there is a different post-processing script after this one in the sequence, that script would now get the wrongly separated layers.

electrocbd replied:
I agree with you, this is what I should do, but I noticed that the provided array merges sometimes two consecutive layers. For example, with https://www.thingiverse.com/thing:301742 (bottom part) using 0.6mm nozzle and default settings on Ultimaker 2+, layers 139 and 140 are merged.
To check it, I just created a very simple script:

```
    def execute(self, data):
        for layer_number, layer in enumerate(data):
            data[layer_number] = ";PPLayer" + str(layer_number) + "\n" + data[layer_number];
        return data;
```

The first three items of the data array contain comments, following items follow correctly the layer changes... until layer 139.

And then, from layer 141 to the end, separation is correct.

Maybe a Cura bug 5 layers before the last one?

Ghostkeeper replied:
Could be a bug there, then!

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,419 @@
+        iflush = 0
+        for i, step in enumerate(layer_steps):
+            if i == 0:
+                current_e = step.step_e
```
So the first step is always considered a travel move now. That doesn't seem correct.

To fix this, it looks like you need to give a starting `GCodeStep` to this function as your current position upon layer switching.

electrocbd replied:
I suppose that layer switching occurs only during a travel move. Knowing the way curaengine works, it should always be the case *except* in spiralize mode, which I never tried.

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,419 @@
+                current_e = step.step_e
+            if current_e == step.step_e:
+                # No extrusion since the previous step, so it is a travel move
+                # Let process steps accumulated into orig_steps,
```
I think this variable was renamed to `orig_seq`.

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,419 @@
+                iflush = i
+                orig_seq = np.empty((0, 2))
+            if step.step == 0 or step.step == 1:
+                orig_seq = np.concatenate([orig_seq, np.array([[step.step_x, step.step_y]])])
```
`np.concatenate` creates a new matrix with every g-code step. Perhaps this can be made more efficient by filling in the entire matrix first and then filtering with `np.where`? Just a suggestion, because I don't know for sure whether this is possible.

The current solution is O(n^2), so it could become very slow with large layers such as skin layers.

electrocbd replied:
I have no clue to make this algorithm faster. It is necessary to know before each extrusion move if material was already deposited at the same Z position (same layer) near the new deposited segment. When drawing on a bitmap, this is very easy: Is there colored pixels around the position where we are going to draw. It is a constant time algorithm.
The problem with the bitmap approach is the amount of memory needed. To have good precision, pixels should be half of the line width. Let say 0,2mm wide.
For a maximum plate dimensions of 20cm x 20cm (and as you already mentioned, some printer are bigger), we need and array of 1000 x 1000 pixels.
Well. It seems doable. Will think of it...

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,419 @@
+        Appends g-code lines to the plugin's returned string
+        """
+        ipos = 0
+        while i < iend:
```
Python still has the construct:

    for i in range(iend):

So a while loop isn't necessary. What's more, you're only using `i` for `layer_steps[i]`, so you could even do:

    for layer_step in layer_steps:

electrocbd replied:
Well... My code is not easily readable... I will try to change again variables names here.
The i variable is a parameter, it is the starting index for the loop, and iend is the end index. The objective is to advance from the last output position (iflush in the calling function) to the position iend.
Your first proposition would restart from the beginning of the layer, and your second proposition would start from the beginning of the layer, and end too far, up to the end of the layer

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,419 @@
+        of continuous material extrusion
+        """
+        if (len(orig_seq) > 2 and
+                ((orig_seq[len(orig_seq) - 1] - orig_seq[0]) ** 2).sum(0) < 0.3 * 0.3):
```
It took quite some figuring out what this does. It seems that it's checking that the Euclidean distance between the last and first element of the sequence is less than 0.3mm. Perhaps you could document that?

And perhaps just using `self.line_width` instead of 0.3mm would be more appropriate.

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,419 @@
+            dist_from_proj = np.sqrt(((projection - step) ** 2).sum(0))
+            if dist_from_proj > 0.001: # Move central point only if points are not aligned
+                modif_seq[i] = (step - (self.stretch / dist_from_proj)
+                                * (projection - step))
```
I wonder how this would mutilate any Triangle Infill :stuck_out_tongue:

electrocbd replied:
The change on the triangle infill is not noticeable. I just tried. When the algorithm moves a point in a triangle, it is for a maximum distance of the stretch distance parameter, about 0.08 mm.

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,419 @@
+        ibeg = 0
+        i = 1
+        iend = 2
+        while i+1 < len(orig_seq):
```
Again, it should be possible to use `for i in range(1, len(orig_seq) - 1):`

Then you wouldn't need to track `good_triangle` any more either, since you can just use `continue` instead of `good_triangle = False` if there's no increment of `i` at the end of the loop.

electrocbd replied:
Ok, I just understood range for loops :-p

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -0,0 +1,419 @@
+            if materialleft and not materialright:
+                modif_seq[ibeg] = modif_seq[ibeg] + xperp * self.stretch
+            elif not materialleft and materialright:
+                modif_seq[ibeg] = modif_seq[ibeg] - xperp * self.stretch
```
This would only check if there is an endpoint nearby. The rest of the sequence could be nearby?

electrocbd replied:
I Don't understand your question. This portion of code checks if, when extruding a segment of material, material was already deposited at the left, at the right, or both (when in a corridor).
If material is already present at the left of the segment, the first point of the segment is slightly moved to the left, in the direction of the wall, and the same for the right.
This operation is done for each point of the sequence.
The objective is to avoid gaps between lines during turns, easily observable without a form of compensation like this algorithm.

Ghostkeeper replied:
My question was about this case.

                          Some other line.
                                  v
    Vertex "A" -> ._________________________________. <- Vertex "B"
                                     /*\ <- Vertex "P"
                                    /   \
                                   /     \ <- Two lines currently being processed
                                  /       \

In this case P is nowhere close to A or B, so this code doesn't detect it. But P is close to the line between A and B.

Ghostkeeper reviewed (changes requested):
The recent changes and refactors made the code a lot more clear. Thanks! Using this new information, I found a few possible bugs that might have to be addressed. I'm not sure if all of them can be fixed (some might be prohibitively expensive to compute).

electrocbd committed:
Use range() for for loops

electrocbd committed:
Test of closed loops uses line width parameter instead of hard coded
value

Ghostkeeper reviewed (approved):
As far as I'm concerned there are still two open problems with this code: The first move after a layer switch, and the detecting of nearby collisions by looking at nearby vertices. But you give some good arguments as to why those may not be a problem so I'm just going to run some tests on this now.

Ghostkeeper commented:
I'm getting a bit of a weird result when applying this to a model with sharp corners:

![image](https://user-images.githubusercontent.com/2448634/27869928-b2b8959a-61a1-11e7-81e1-b8dbab522b37.png)

For this test I set the distance to 1mm in order to be able to see the result.
For some corners it did offset the vertex, but for others it didn't. Do you know why?

Ghostkeeper commented:
With a simple cube it becomes apparent that the result is deterministic:
![image](https://user-images.githubusercontent.com/2448634/27870118-4b5cb3bc-61a2-11e7-984e-6f81c07cf907.png)
The separation between the part with skin and the part with infill is clear. But it still seems to be somewhat random: For the parts where there is skin it offset the vertex in the front-right corner (the other 3 are untouched). For the parts where there is infill it offset the vertex in the back-left corner. The front-left corner is always inset and the back-right corner never. The weirdness at the brim is where the brim polygons start and end (there are travel moves there to get to the next brim line).

Ghostkeeper commented:
A cylindrical ring is the intended target of this feature, but that also produces some weird phenomena in the walls.
![image](https://user-images.githubusercontent.com/2448634/27870635-a8f65b6c-61a3-11e7-9400-207169b65505.png)

I think the way you're deciding whether or not to offset is still off.

Ghostkeeper commented:
When z-hops are required, you get a result in the command line like this:

    2017-07-05 17:13:17,954 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [89]: Post stretch with line width=0.35mm and stretch=0.08mm
    2017-07-05 17:13:17,954 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.000 27 steps
    2017-07-05 17:13:17,955 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.000 8 steps
    2017-07-05 17:13:17,958 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.270 119 steps
    2017-07-05 17:13:17,989 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.270 5 steps
    2017-07-05 17:13:17,990 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.270 95 steps
    2017-07-05 17:13:18,000 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.270 5 steps
    2017-07-05 17:13:18,030 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.270 1875 steps
    2017-07-05 17:13:18,434 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.270 2 steps
    2017-07-05 17:13:18,436 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.270 194 steps
    2017-07-05 17:13:18,456 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.270 8 steps
    2017-07-05 17:13:18,456 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.370 4 steps
    2017-07-05 17:13:18,457 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.370 95 steps
    2017-07-05 17:13:18,465 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.370 5 steps
    2017-07-05 17:13:18,479 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.370 1312 steps
    2017-07-05 17:13:18,693 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.370 5 steps
    2017-07-05 17:13:18,700 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.370 776 steps
    2017-07-05 17:13:18,801 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.370 7 steps
    2017-07-05 17:13:18,801 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.470 4 steps
    2017-07-05 17:13:18,802 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.470 95 steps
    2017-07-05 17:13:18,808 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.470 5 steps
    2017-07-05 17:13:18,829 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.470 1875 steps
    2017-07-05 17:13:19,174 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.470 2 steps
    2017-07-05 17:13:19,176 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.470 193 steps
    2017-07-05 17:13:19,197 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.470 5 steps
    2017-07-05 17:13:19,197 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.570 4 steps
    2017-07-05 17:13:19,198 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.570 95 steps
    2017-07-05 17:13:19,207 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.570 5 steps
    2017-07-05 17:13:19,222 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.570 1312 steps
    2017-07-05 17:13:19,432 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.570 5 steps
    2017-07-05 17:13:19,439 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.570 776 steps
    2017-07-05 17:13:19,535 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.570 4 steps
    2017-07-05 17:13:19,536 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.670 4 steps
    2017-07-05 17:13:19,537 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.670 95 steps
    2017-07-05 17:13:19,543 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.670 5 steps
    2017-07-05 17:13:19,564 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.670 1875 steps
    2017-07-05 17:13:19,910 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.670 2 steps
    2017-07-05 17:13:19,912 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.670 193 steps
    2017-07-05 17:13:19,934 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.670 4 steps
    2017-07-05 17:13:19,935 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.770 4 steps
    2017-07-05 17:13:19,935 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.770 95 steps
    2017-07-05 17:13:19,946 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.770 5 steps
    2017-07-05 17:13:19,960 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.770 1312 steps
    2017-07-05 17:13:20,168 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.770 5 steps
    2017-07-05 17:13:20,176 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.770 776 steps
    2017-07-05 17:13:20,270 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.770 4 steps
    2017-07-05 17:13:20,270 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.870 4 steps
    2017-07-05 17:13:20,271 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.870 95 steps
    2017-07-05 17:13:20,277 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.870 5 steps
    2017-07-05 17:13:20,297 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 0.870 1875 steps
    2017-07-05 17:13:20,639 - DEBUG - PostProcessingPlugin.scripts.Stretch.execute [118]: Layer Z 2.870 2 steps

That appears to me like the assumption that the first move after a Z change is always a travel move is not correct then. Z hops are enabled by default in quite a lot of our profiles.

Ghostkeeper commented:
Also with a more normal offset distance (0.08mm, the default) the result is a bit random (but less visible):

![image](https://user-images.githubusercontent.com/2448634/27871788-0e8287dc-61a7-11e7-8a50-8ca00104c066.png)

electrocbd commented:
Z-Hop seem to be the most obvious issue... If a correctly understood the "contract" each plugin should follow, the input is an array with one element per layer, and the output should also be an array with one element per layer.
I will try to fix this soon.

electrocbd committed:
Layer really changes only when material is extruded (Z-Hop issue)

electrocbd committed:
Output array allow to chain plugins

electrocbd commented:
I just understood what you meant when you were talking about "your assumption is that the first movement is a travel move". I am currently skipping the compensation for the first segment of each sequence :-(
I will fix it soon

electrocbd committed:
wideCircle algorithm debug

electrocbd commented:
Just a reminder: My last commit (I did several weeks ago) should fix issues.

Ghostkeeper commented:
Ah, sorry. I saw that commit but dismissed it as debugging. Code looks all right; it doesn't do anything much stranger than what the code already did.

We'll test it again. I'm marking the issue in our internal tracker.

Ghostkeeper commented:
I'm still seeing the same problem on the arrow model:
![image](https://user-images.githubusercontent.com/2448634/29029434-f5f854ee-7b87-11e7-8874-f79e0f43428e.png)

This also exhibits on a simple cylinder:
![image](https://user-images.githubusercontent.com/2448634/29029871-69f20402-7b89-11e7-8559-0b56b5ae184e.png)

And the hollow cylinder:
![image](https://user-images.githubusercontent.com/2448634/29030484-7061677c-7b8b-11e7-8200-838bc3e77127.png)


For some reason sometimes the lines are offset but sometimes not. You can see this in the extreme because I've set the offset in the script to 1mm. As a result the inner walls sometimes get moved outside of the outer walls. Perhaps this has to do with adjacent infill lines.

I'd expect the outer wall to always be offset. That's the most important thing. Right now it seems to never offset. The inner wall is sometimes offset, causing the inner walls to be moved closer to the outer walls.

electrocbd commented:
What you notice for the cylinder is a consequence of the "pushwall" algorithm. When the lines of the triangle infill touch the inner walls, the algorithm goes in the direction of the contact.
This allows to take into account the interaction between segments of extruded plastic when they touch each other.
Of course, it is possible to disable it, and to have a better visual aspect on the 3D view, but I try to have a better visual aspect on the effectively printed object, which is not exactly what the 3D view supposed it would be...
Well, anyway, this plugin's objective is better dimensional accuracy, not better surface finish ;-)

electrocbd commented:
If you use a more reasonable stretch factor (let say 0.2mm), you will see that the affected wall is the inner wall, not the outer wall. With normal settings, the outer wall should not by influenced by the internal infill.
![screenshot_20170807_201710](https://user-images.githubusercontent.com/6553898/29039931-93f4f75e-7bad-11e7-9a13-ed5c54a8ac74.png)

electrocbd commented:
You expect an increase in external diameters, but I noticed issues with _internal_ diameters, using a caliper and calibrated cylinders to measure effective inner and outer diameters of printer cylinders.
When the printer extrudes the outer wall, generally, this outer wall is pushed in the outside direction by the already extruded inner wall and, most important, _it depends on the order of operations_. If the outer wall is extruded before the inner wall (which is of course not a very good idea), its diameter is narrower.
The 3D view doesn't help much to see the effect of the order of the extrusions (inner walls printed from outside to inside or the opposite, etc.)

Ghostkeeper commented:
What I see though is that on the outside also some vertices get their diameter modified and some vertices don't get moved. The result is an irregular surface.

The same thing happens on the inside corners.

electrocbd commented:
Can you share the STL file and the settings you used, to help me check the issue?

Ghostkeeper commented:
Rename this file to `StretchScript.curaproject.3mf` and open it with Cura: [StretchScript.curaproject.zip](https://github.com/nallath/PostProcessingPlugin/files/1266806/StretchScript.curaproject.zip)
Then add the Stretch script and set the stretch distance to 0.2mm.

You'll see that some corners on the outside don't get moved while other corners do get moved.
![image](https://user-images.githubusercontent.com/2448634/29920330-507186fc-8e4d-11e7-9123-107cb5274ce1.png)

electrocbd committed:
Split pushwall and widecircle parameters to enable to fine tune the effect
Very short travel moves do not break extrusion sequences

electrocbd commented:
I just pushed a new version of the plugin. The main issue I found using your test case (which I applied like a charm, thanks 🥇 ) was very short moves without extrusions. CuraEngine seem to create sometimes very short moves, less than a quarter of the nozzle size, without extrusion (the E parameter do not change). As a consequence, the plugin breaks its sequence, and is unable to find a closed loop to apply the `wideCircle` algorithm.
I fixed it, ignoring travel moves less than half the line extrusion width. It seems to enhance the "stability" of the algorithm.
I also added a new parameter, to be able to play separately with the `wideCircle / WideTurn` and the `pushWall` algorithm. Setting 0 for a parameter disables the corresponding algorithm

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -72,22 +72,27 @@ class Stretcher():
+        self.wc_stretch = wc_stretch
+        self.pw_stretch = pw_stretch
+        if self.pw_stretch > line_width / 4:
+            self.pw_stretch = line_width / 4 # Limit value of pushwall stretch distance
```
There is the `max(...)` function for this ;) Oh well...

Ghostkeeper commented on scripts/Stretch.py:
```diff
@@ -72,22 +72,27 @@ class Stretcher():
-                   + "mm and stretch=" + str(self.stretch)+ "mm")
+        Logger.log("d", "Post stretch with line width = " + str(self.line_width)
+                   + "mm wide circle stretch = " + str(self.wc_stretch)+ "mm"
+                   + "and push wall stretch = " + str(self.pw_stretch) + "mm")
```
For this sort of thing we have formatting, for instance:

    Logger.log("d", "Post stretch with line width = {line_width}mm wide circle stretch = {wc_stretch}mm and push wall stretch = {pw_stretch}mm",
        line_width = self.line_width,
        wc_stretch = self.wc_stretch,
        pw_stretch = self.pw_stretch)

Ghostkeeper commented:
It has improved. There's no irregularities in the inside walls any more due to the infill. It's still sometimes messy though, like where the two circles meet here:
![image](https://user-images.githubusercontent.com/2448634/30272674-f18aa84c-96f5-11e7-837d-41086cf846b4.png)
Can something be done about that?

I'm willing to merge this now as-is, but if you know of something that could help... Perhaps the normal vector detection doesn't work properly in cases such as this. Note that in the original g-code, one line width is significantly smaller than the other because of overlap compensation.

Outside corners are sometimes still irregular (like in the arrow examples I posted earlier).

electrocbd committed:
Enhance wideCircle algorithm stability
Don't cancel wideCircle correction when surrounded by walls
Change default stretch distance from 0.08 to 0.1mm

electrocbd commented:
I just changed two internal parameters, to enhance wideCircle algorithm stability. The system use now wider triangles to estimate directions of turns. This lowers chance for choosing the wrong direction for the correction of the movement.
Resulting gcode is cleaner, but I haven't tried (yet) to print it to check the effective accuracy of hole sizes.
So far, I have no clue to do better.

Ghostkeeper commented:
Let's merge this now then and make incremental improvements later. We'll see what our QA team has to say about it.

Ghostkeeper commented:
This has been merged: https://github.com/nallath/PostProcessingPlugin/commit/7603763d63e579e3ed210af2be5922a49d215ce3

For some reason, Github doesn't show it as being merged though, so I'll close this PR manually.
